workload: retry errors when tolerate-errors is set

andrewbaptist opened the pull request:
Previously if tolerate errors was set, the workload would log and drop errors. Now it will retry the errors. This more closely simulates how clients would use our system.

Informs: #123304

Epic: none

Release note: None

Commit:
workload: retry errors when tolerate-errors is set

Previously if tolerate errors was set, the workload would log and drop
errors. Now it will retry the errors. This more closely simulates how
clients would use our system.

Informs: #123304

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123305)
<!-- Reviewable:end -->

Commit:
workload: remove unnecessary nil check

Release note: None

Commit:
workload: allow a catchup burst

Previously the limiter would respond to a cluster slowdown by slowing
down its rate if --max-rate was set, and it would never recover the
missed requests. With this change it will save the missed requests in
the limiters burst and issue requests faster after the cluster can go
faster. This turns the --max-rate into a maximum over a longer interval
rather than always starting from now.

As part of this change the workload is ramped up more consistently if
max rate is set by both ramping up the number of workers as well as the
number of requests per second each worker makes. Previously --max-rate
and --ramp did not play well together.

Release note: None
